accounts/abi/bind/bindv2.go

```diff
@@ -0,0 +1,311 @@
+	normalized.Name = normalizedName
+	normalized.Inputs = cb.normalizeErrorOrEventFields(original.Inputs)
+	cb.errors[original.Name] = &tmplError{Original: original, Normalized: normalized}
+	return nil
+}
+
+func parseLibraryDeps(unlinkedCode string) (res []string) {
+	reMatchSpecificPattern, err := regexp.Compile(`__\$([a-f0-9]+)\$__`)
+	if err != nil {
+		panic(err)
+	}
+	for _, match := range reMatchSpecificPattern.FindAllStringSubmatch(unlinkedCode, -1) {
+		res = append(res, match[1])
+	}
+	return res
+}
+
+// BindV2 generates a Go wrapper around a contract ABI. This wrapper isn't meant
+// to be used as is in client code, but rather as an intermediate struct which
+// enforces compile time type safety and naming convention as opposed to having to
+// manually maintain hard coded strings that break on runtime.
+func BindV2(types []string, abis []string, bytecodes []string, pkg string, libs map[string]string, aliases map[string]string) (string, error) {
+	b := binder{
+		contracts: make(map[string]*tmplContractV2),
+		structs:   make(map[string]*tmplStruct),
+		aliases:   aliases,
+	}
+	for i := 0; i < len(types); i++ {
+		// Parse the actual ABI to generate the binding for
+		evmABI, err := abi.JSON(strings.NewReader(abis[i]))
+		if err != nil {
+			return "", err
+		}
+
+		for _, input := range evmABI.Constructor.Inputs {
+			if hasStruct(input.Type) {
+				bindStructType(input.Type, b.structs)
+			}
+		}
+
+		cb := contractBinder{
+			binder: &b,
+			calls:  make(map[string]*tmplMethod),
+			events: make(map[string]*tmplEvent),
+			errors: make(map[string]*tmplError),
+
+			callIdentifiers:  make(map[string]bool),
+			errorIdentifiers: make(map[string]bool),
+			eventIdentifiers: make(map[string]bool),
+		}
+		for _, original := range evmABI.Methods {
+			if err := cb.bindMethod(original); err != nil {
+				return "", err
+			}
+		}
+
+		for _, original := range evmABI.Events {
+			if err := cb.bindEvent(original); err != nil {
+				return "", err
+			}
+		}
+		for _, original := range evmABI.Errors {
+			if err := cb.bindError(original); err != nil {
+				return "", err
+			}
+		}
+		// Strip any whitespace from the JSON ABI
+		strippedABI := strings.Map(func(r rune) rune {
+			if unicode.IsSpace(r) {
+				return -1
+			}
+			return r
+		}, abis[i])
+		// replace this with a method call to cb (name it BoundContract()?)
+		b.contracts[types[i]] = &tmplContractV2{
+			Type:        capitalise(types[i]),
+			InputABI:    strings.ReplaceAll(strippedABI, "\"", "\\\""),
+			InputBin:    strings.TrimPrefix(strings.TrimSpace(bytecodes[i]), "0x"),
+			Constructor: evmABI.Constructor,
+			Calls:       cb.calls,
+			Events:      cb.events,
+			Errors:      cb.errors,
+			Libraries:   make(map[string]string),
+		}
+	}
+
+	invertedLibs := make(map[string]string)
+	for pattern, name := range libs {
+		invertedLibs[name] = pattern
+	}
+
+	data := tmplDataV2{
+		Package:   pkg,
+		Contracts: b.contracts,
+		Libraries: invertedLibs,
```

Comment:
Re `Libraries` here. So, in the original `v1` you have modified it to be:
```golang
type tmplData struct {
	Package      string                   // Name of the package to place the generated file in
	Contracts    map[string]*tmplContract // List of contracts to generate into this file
	Libraries    map[string]string        // Map the bytecode's link pattern to the library name
	InvertedLibs map[string]string        // map of the contract's name to the link pattern
	Structs      map[string]*tmplStruct   // Contract struct type definitions
}
```
But why is that modified in this PR? 

-----

Anyway, originally, it was 
```golang
type tmplData struct {
	Package   string                   // Name of the package to place the generated file in
	Contracts map[string]*tmplContract // List of contracts to generate into this file
	Libraries map[string]string        // Map the bytecode's link pattern to the library name
	Structs   map[string]*tmplStruct   // Contract struct type definitions
}
```
And the new version has 
```golang
type tmplDataV2 struct {
	Package   string                     // Name of the package to place the generated file in
	Contracts map[string]*tmplContractV2 // List of contracts to generate into this file
	Libraries map[string]string          // Map of the contract's name to link pattern
	Structs   map[string]*tmplStruct     // Contract struct type definitions
}
```
Since both are string-string mappings, it's a bit hard to know for sure, but IIUC, you have inverted the mapping. Isn't that a bit user-hostile?  

It's a bit odd.. The `v2` is essentially identical to the old `v1`, from a struct-perspective, but the semantics of one of the fields has been inverted. What is the reason for inverting it? 


Moreover, you do 
```golang
	for pattern, name := range libs {
		invertedLibs[name] = pattern
	}
``` 
But if there are any `pattern` (that is, contracts?) which have the same libraries, then they will overwrite eachother. 

```
         A
       /    \      
    B        D
      \     /
        E
```
Then I think the entry for `A` will list _either_ `B` or `D`.

Reply:
The changes to the original structs are leftover cruft that is no longer wanted.  The `invertedLibs` is cruft.